README.md
# The Game of Patterns

## Installing

You can install the game by simply running:

```bash
make
make install
```

Keep in mind that you may have to run these as root.

To run the game, run `gameofpatterns`.

To uninstall the game, run `make uninstall` or `make clean`.

## How to Play

Complete the numeric pattern you are randomly given. You are prompted
with 4 items, and then must complete the pattern up to 10.

## Features

* There are currently 10 patterns to solve (with more being actively added)
* The game doesn't repeat patterns until you have completed them all (as of v0.2.0)
* The pattern is auto-solved if you make a mistake

## A Note about the Versions

I released versions 0.1.0 and 0.2.0 alongside v0.3.0 to keep everything in order when I tranferred the
code to GitHub. Note that these two versions have no Makefile and are only available in a tarball format
(no zip archive included). You can use these versions, but I suggest using the latest version.


CHANGELOG.md
# Game of Patterns - Changelog

## v0.3.0 - Released 2020-11-13

* First release of the software on GitHub

* **Added**
  * Added an EditorConfig file
  * Added 2 new patterns

* **Changed**
  * Refactored the pattern solving method
  * You can now build/install using `make`/`make install`
  * Re-formatted the changelog

## v0.2.0 - Released 2020-07-30

* **Added**
  * Added an anti-repeat feature
  * Added full changelog

* **Changed**
  * Re-formatted some of the in-game displays
  * Updated the README to make it more readable and user-friendly

* **Fixed**
  * Fixed three broken patterns
  * You can now complete the game

## v0.1.0 - Released 2020-06-25

* Initial release of Game of Patterns


src/main.cpp
//
// FILENAME: main.cpp | GameOfPatterns
// DESCRIPTION: The source file for GameOfPatterns
// CREATED: 2020-06-22 @ 3:27 AM
// COPYRIGHT: Copyright (c) 2020 by Ryan Smith <rysmith2113@gmail.com>
//

#include <iostream>
#include <utility>
#include <cstdlib>
#include <csignal>

#include "patterngen.hpp"
#include "solver.hpp"

using std::cout;
using std::cin;
using std::endl;
using std::signal;

void sigHandler(int signo);

bool dispTitle(); // Returns `true` if the user wants to play
bool rePrompt(); // Prompts the user to play again 

int main() {
	PatternGen* pg = new PatternGen;
	Solver* solver = new Solver;

	// Register a signal handler
	signal(SIGINT, sigHandler); // CTRL-C/CTRL-D

	system("clear"); // Clear the console

	bool willPlay = dispTitle(); // Run the title sequence

	if (!willPlay) {
		cout << "Quitting...\n";
		return 0;
	}

	game:
		system("clear");

		pair<vector<int>, ushort> patternData = pg->GetPattern();

		bool completed = false, incorrect = false;

		int userNext; // The next item the user enters
		int itemsSolved = 0;

		// Split the pair
		vector<int> pattern = patternData.first;
		ushort patternID = patternData.second;

		system("clear");

		while (!completed && !incorrect) {
			cout << "\033[1;31m"
				<< "Here is your pattern:"
				<< "\033[0m\n";

			// Iterate through the pattern and disply it
			for (unsigned int i = 0; i < pattern.size(); i++) {
				cout << pattern[i] << ", ";
			}

			// Next item "?"
			cout << "\033[;40m\033[7;37m"
				<< "?"
				<< "\033[0m\n";

			bool good = false;

			while (!good) {
				cin.clear(); // Reset cin flags

				cout << "Next item: ";
				cin.ignore();
				cin >> userNext;

				if (!cin.good()) {
					cout << endl;
				}

				good = cin.good();
			}

			cin.clear();

			int next = solver->GenNext(pattern, patternID);

			system("clear"); // Clear for formatting

			if (next == userNext) {
				cout << "\033[1;32m"
				<< "Correct!"
				<< "\033[0m\n\n";

				pattern.push_back(userNext); // Add the next item to the pattern
				itemsSolved++;
			}
			else {
				cout << "\033[1;31m"
				<< "Incorrect!"
				<< "\033[0m\n\n";

				incorrect = true;
			}

			// Check to see if the pattern was completed
			if (itemsSolved == 4) {
				completed = true;
			}
		}

		if (completed) {
			cout << "\033[1;34m"
				<< "You finished the pattern!"
				<< "\033[0m\n\n";
		}
		else {
			cout << "\033[1;34m"
				<< "You got " << itemsSolved << "/6 additional items in the pattern."
				<< "\033[0m\n\n";

			cout << "\033[1;34m"
			<< "Final pattern:"
			<< "\033[0m\n\n";

			for (unsigned int i = 0; i < pattern.size(); i++) {
				cout << pattern[i] << ", ";
			}

			cout << "\033[;31m" << userNext << "\033[0m";
			cout << "\033[;32m" << " (" << solver->GenNext(pattern, patternID) << ")" << endl;
			cout << endl << endl;
		}

	if ((rePrompt()) && (!pg->UsedAll())) {
		goto game;
	}

	return 0;
}

void sigHandler(int signo) {
	cout << "\nQuitting...\n";
	exit(signo);
}

bool dispTitle() {
	cout << "This is GameOfPatterns v0.3.0.\tCopyright (c) 2020 by Ryan Smith.\n\n";

	// Title - red with bold and underline
	cout << "\033[4;31m\033[1;31m"
		 << "Welcome to the Game of Patterns"
		 << "\033[0m\n\n";

	// Prompt - Blue text with bold; "y/n" - green/red (respectively) with bold
	cout << "\033[1;34m"
		 << "Would you like to play ("
		 << "\033[1;32m" << "y"
		 << "\033[1;34m" << "/"
		 << "\033[1;31m" << "n"
		 << "\033[1;34m" << ")?"
		 << "\033[0m\n\n";

	char answer = ' ';

	do {
		cout << ">>> ";
		cin >> answer;
	} while ((answer == ' ') || ((answer != 'y') && (answer != 'n')));

	return (answer == 'y')? true : false; // Whether or not the user wants to play
}

bool rePrompt() {
	// Prompt - Blue with bold; "y/n" - green/red (respectively) with bold
	cout << "\033[1;34m"
		 << "Would you like to play again ("
		 << "\033[1;32m" << "y"
		 << "\033[1;34m" << "/"
		 << "\033[1;31m" << "n"
		 << "\033[1;34m" << ")?"
		 << "\033[0m\n\n";

	char answer = ' ';

	do {
		cout << ">>> ";
		cin >> answer;
	} while ((answer == ' ') || ((answer != 'y') && (answer != 'n')));

	return (answer == 'y')? true : false; // Whether or not the user wants to play
}


src/solver.hpp
//
// FILENAME: solver.hpp | GameOfPatterns
// DESCRIPTION: The solver class for calculating the next number in apattern or completeting the pattern
// CREATED: 2020-06-22 @ 2:16 AM
// COPYRIGHT: Copyright (c) 2020 by Ryan Smith <rysmith2113@gmail.com>
//

#ifndef SOLVER_HPP
#define SOLVER_HPP

#include <cmath>
#include <vector>
#include <map>

using std::vector;
using std::map;

class Solver {
public:
	int GenNext(vector<int> pattern, ushort id) const;
	vector<int> Solve(vector<int> pattern, ushort id);
};

int Solver::GenNext(vector<int> pattern, ushort id) const {
	// Generate the next element in a given pattern
	map<int, int> solutions = {
		{0, pow(pattern.size() + 1, 2)},						// Square the position of the next item in the sequence (5th element is 25)
		{1, (pattern.back() + pattern.at(pattern.size() - 2))},	// Fibonacci Sequence
		{2, pow(pattern.size() + 1, 3)},						// Cube the position of the next item in the sequence (5th element is 125)
		{3, (pow(pattern.back(), 2) - 1)},						// Square the next element then subtract 1
		{
			4, ((pattern.at(pattern.size() - 2) - 3) == pattern.back())? \
			(pattern.back() + 1) /* Add 1 if 3 was subtracted last */ : \
			(pattern.back() - 3) /* Subtract 3 if 1 was added last */
		},														// Subtract 3, add 1
		{5, pattern.back() * 3},								// Multiply by 3
		{6, pow(10, pattern.size() + 1)},						// Raise the next element's position to the 10th power
		{7, pattern.back() * pattern.size()},					// Multiply the last element by the position
		{8, pattern.back() + 17},								// Add 17
		{9, pattern.back() * 8},								// Multiply by 8
	};
	
	return solutions[id];
}

vector<int> Solver::Solve(vector<int> pattern, ushort id) {
	vector<int> solvedPattern;

	// The pattern after a new element is added; passed into this->GenNext(vector<int>, ushort)
	vector<int> incrPattern = pattern;

	// Solve the pattern (complete it with the 6 remaining items)
	for (unsigned int i = pattern.size() /* 4 */; i < 10 /* Future size */; i++) {
		incrPattern.push_back(this->GenNext(incrPattern, id));
	}

	solvedPattern = incrPattern;

	return solvedPattern;
}

#endif // SOLVER_HPP

src/patterngen.hpp
//
// FILENAME: cache.py | GameOfPatterns
// DESCRIPTION: The PatternGen class for choosing a pattern to prompt the player with
// CREATED: 2020-06-22 @ 3:24 AM
// COPYRIGHT: Copyright (c) 2020 by Ryan Smith <rysmith2113@gmail.com>
//

#ifndef PATTERNGEN_HPP
#define PATTERNDEF_HPP

#include <string>
#include <vector>
#include <utility>
#include <algorithm>

using std::string;
using std::vector;
using std::pair;
using std::find;

class PatternGen {
public:
	// Return the pattern and its ID number (position in the vector)
	pair<vector<int>, ushort> GetPattern();

	bool UsedAll(); // Check if all of the patterns have been used
 
private:
	// Store the patterns
	vector<vector<int>> m_Patterns = {
		{1, 4, 9, 16},   	    // Square the position of the next number in the sequence (next: 5 -> 25, 6 -> 36 ...)
		{1, 1, 2, 3},    	    // Fibonacci Sequence (next: 5, 8, 13...)
		{1, 8, 27, 64},  	    // Cube the position of the next number in the sequence (next: 5 -> 125, 6 -> 216 ...)
		{2, 3, 8, 63},   	    // Square the next element then subtract 1 (next: 63^2 - 1 ...)
		{10, 7, 8, 5},   	    // Subtract 3, add 1 (next: 6, 3, 4...)
		{7, 21, 63, 189}, 	    // Multiply by 3 (next: 567, 1701, 5103...)
		{10, 100, 1000, 10000}, // Raise 10 to the power of the position of the next number in the sequence (next: 10^5, 10^6)
		{1, 2, 6, 24},        	// Multiply by an incremental value of one 
		{17, 34, 51, 68},		// Count by 17s (next: 85, 102, 119...)
		{5, 40, 320, 2560},		// Multiply by 8 (next: 20544, 164325...)
	};

	vector<ushort> m_UsedIDs; // Used to prevent repeat patterns
};

pair<vector<int>, ushort> PatternGen::GetPattern() {
	pair<vector<int>, ushort> returnPair;
	
	vector<ushort>::iterator iter;
	
	srand(static_cast<unsigned int>(time(0)));
	int randomNum = rand() % m_Patterns.size();
	
	// Randomly select a pattern to be used
	if (!this->m_UsedIDs.empty()) {
		do {
			// Pick a random number
			srand(static_cast<unsigned int>(time(0)));
			randomNum = rand() % m_Patterns.size();
			
			iter = find(this->m_UsedIDs.begin(), this->m_UsedIDs.end(), randomNum); // Check for repeats
		} while (iter != this->m_UsedIDs.end()); // Will loop until an ID is not found in the vector
	}

	this->m_UsedIDs.push_back(randomNum); // Record the pattern ID

	returnPair.first = m_Patterns[randomNum]; // The pattern
	returnPair.second = static_cast<ushort>(randomNum); // The pattern ID

	return returnPair;
}

bool PatternGen::UsedAll() {
	return (this->m_UsedIDs.size() == this->m_Patterns.size());
}

#endif // PATERNGEN_HPP